File: compiler/bench/pell.c
// expected: 0
// Pell numbers - P(n) = 2*P(n-1) + P(n-2)
#include <stdio.h>
#include <stdint.h>

long pell(int n) {
    if (n == 0) return 0;
    if (n == 1) return 1;
    long a = 0, b = 1;
    for (int i = 1; i < n; i++) {
        long t = 2 * b + a;
        a = b;
        b = t;
    }
    return b;
}

int main(void) {
    long result = 0;
    for (int i = 0; i < 5000000; i++) {
        result ^= pell(i % 32);
    }
    printf("%ld\n", result);
    return 0;
}

File: compiler/bench/pell.fs
\ expected: 0
\ Pell numbers - P(n) = 2*P(n-1) + P(n-2)

: pell ( n -- P_n )
  dup 0= if exit then
  dup 1 = if exit then
  0 1                     \ a=P0=0, b=P1=1 (stack: n a b)
  rot 1 do                \ loop from 1 to n-1
    tuck 2 * +            \ (a b) -> (b a b) -> (b a 2b) -> (b 2b+a)
  loop nip ;

: main
  0
  5000000 0 do
    i 32 mod pell xor
  loop
  . cr ;
